Support refresh tokens

jwoudenberg opened the pull request:
Resolves #12

This adds supports for refresh tokens to the OAuth2 middleware.

It also splits out a dedicated OpenID Connect middleware, because it's session invalidation logic is slightly different. Up to this point I've been using the oauth2 middleware for an OpenID Connect flow, making use of the fact that we can fetch an ID token by sending an `idtoken` scope and then read that id token out of the session. To add refresh token support we need to be able to figure out when a session has expired though, and here the OAuth2 and OpenID implementations diverge:

- For the OAuth2 flow we check the expire time of the access token.
- For the OpenID Connect flow we check the expire time of the id token.

I think existing users of the OAuth2 provider probably wouldn't want their sessions to become invalidated because of some id token they might not use or know about. Conversely those looking to rely on ID tokens might not need an access token or be interested in its expiry.

I also added some test for old and new functionality of the OAuth2 and OpenID Connect providers. Happy to look into setting up some CI (maybe based on Travis?) if that seems useful!

Please let me know what you think!

jwoudenberg committed:
Allow providers to validate and refresh sessions

This change allows authentication providers to implement two new types
of behavior: expiring a session and refreshing a session. Both changes
were made with the intent of supporting upcoming changes in the OAuth2
providers, but might find use in other types of authentication providers
as well.

The use case for expiration is that OAuth2 authentication results in an
access token that often has an expiry date. Using the access token
to make requests beyond that date will result in an error response. The
auth middleware has the opportunity to detect an invalid token early and
see the user replaces it, pre-empting a failed request later and so
resulting in a better user experience.

The use case for refreshing a session is that OAuth2 also provides a
mechanism for obtaining a new access token without involving the user,
by using a refresh token. When we obbtain a new access token this way we
do need to update the session to store it, and this commit makes that
possible.

To maintain backwards compatibility we add a default behavior to all
existing providers that makes them behave like before this change. The
default behavior will never cause sessions to be expired or refreshed.

jwoudenberg committed:
OAuth2 middleware supports refresh tokens

jwoudenberg committed:
Extract implementation of OAuth2 provider

This is preparation for creation of a new Open ID Connect provider,
which will now be able to reuse most of the logic for the OAuth2
provider.

jwoudenberg committed:
Add tasty test runner

This is preparation for adding additional tests. Our existing tests are
hedgehog property tests, the next ones aren't going to be. The tasty
wrapper will allow us to run these different types of tests as a single
suite.

jwoudenberg committed:
Add tests for OAuth2 middleware

jwoudenberg committed:
Add OpenID Connect provider

jwoudenberg committed:
Add tests for OpenIDConnect provider

jwoudenberg committed:
Clean up test logic of OAuth2 and OpenID providers

jwoudenberg committed:
Make oauth2 and oidc test suites more similar

jwoudenberg committed:
Update CHANGELOG

I'm not sure if this is the definitive version. We can change it if not.

jwoudenberg committed:
Move metadata into the `Internal` module

It's not part of the public API. We could expose the type, but there's
(at the moment) nothing you would be able to do with it. Better for it
not to show up in documentation then.

jwoudenberg committed:
Rename OpenIDConnect to OIDC

It's a common shorthand for OpenIDConnect and makes for a shorter module
name.

jwoudenberg committed:
Tweak documentation

jwoudenberg committed:
`discover` function takes a text

That's one less library (uri-bytestring) a user needs needs to install
and import to be able to use this middleware.

snoyberg commented:
You interested in being a comaintainer of this package?

jwoudenberg commented:
Sure, happy to help! Thank you for offering!

snoyberg commented:
Invite sent here on Github. What's your Hackage username?

jwoudenberg commented:
It's `JasperWoudenberg`

snoyberg commented:
@lehins can you grant @jwoudenberg (and myself if convenient) maintainer access on Hackage?

lehins commented:
Done. You are both maintainers of this package on hackage.

jwoudenberg committed:
Support older versions of `jose` packages

jwoudenberg committed:
Fix decoding of oidc discovery data

jwoudenberg commented:
Quick update on my radio silence here: I've been wanting to try this out on a local project before calling it 'done'. I've recently started doing that and found a couple of small issues. Working on fixing those, and then I'll get back here!

jwoudenberg commented:
This seems to work 🎉. @lehins would you be interested in reviewing? I realize that as a maintainer I'm able to move this forward myself, but I'm not sure about the PR etiquette in this repo!

lehins commented on src/Network/Wai/Middleware/Auth/OIDC.hs:
```diff
@@ -0,0 +1,281 @@
+-- library. Pre-0.8.x `JWT.string` produces a `String`. Post-0.8.x it produces a
+-- `Text`. This type class allows us to support both.
+--
+-- We can drop this once we no longer wish to support `jose` versions 0.7.x and
```
I think it is ok to drop the support for jose-0.7.x right away in favor of reduced complexity.
Moreover, jose-0.8.0 has been out for a year and a half and is already in 3 stackage lts snapshots.

jwoudenberg replied:
Makes sense!

lehins commented on src/Network/Wai/Middleware/Auth/OIDC.hs:
```diff
@@ -0,0 +1,281 @@
+      -- | The HTTP manager to use. Defaults to the global manager.
+      --
+      -- @since 0.2.3.0
+      , oidcManager :: Manager
```
One of uses for the auth providers is ability to read their configuration from a yaml file. Sticking a manager into the config will make it impossible to give it a FromJSON instance. See Google+OAuth2 for example: 
https://github.com/fpco/wai-middleware-auth/blob/c9544ab9ad631c612f7940340c3d9c56a0c219dc/src/Network/Wai/Middleware/Auth/OAuth2/Google.hs#L72-L84

jwoudenberg replied:
That makes sense! Reason I put the `Manager` in there is that I'd like to be able to use a customized version of the manager in our application (one that adds tracing to outgoing requests).

I think both use cases can be served by letting the `FromJSON` instance always insert the global manager. Going to give that a try!

jwoudenberg replied:
Fixed by making `oidcManager` a `Maybe Manager`.

lehins commented on src/Network/Wai/Middleware/Auth/OIDC.hs:
```diff
@@ -0,0 +1,281 @@
+      --
+      -- @since 0.2.3.0
+      , oidcAllowedSkew :: Clock.NominalDiffTime
+      }
```
We need to be able to grab the config from a yaml file, which is required for `wai-auth` binary. If we are to add OpenID Connect, we need a `FromJSON` instance for this data type.

jwoudenberg replied:
Cool, will do this!

jwoudenberg replied:
Good news! Looks like Aeson has bundled a `FromJSON NominalDiffTime` instance for a couple of years, so there's nothing for us to do here!

lehins commented on src/Network/Wai/Middleware/Auth/OAuth2.hs:
```diff
@@ -104,46 +87,58 @@ instance AuthProvider OAuth2 where
+      Right tokens -> do
+        CTime now <- epochTime
+        if tokenExpired user now tokens then do
+          rRes <- refreshTokens tokens man oauth2
```
It is better to move the lines above here since neither `oauth2` nor `man` are needed in the `else` case:

```haskell
    authEndpointURI <- parseAbsoluteURI oa2AuthorizeEndpoint
    accessTokenEndpointURI <- parseAbsoluteURI oa2AccessTokenEndpoint
    let oauth2 =
          OA2.OAuth2
          { oauthClientId = getClientId oa2ClientId
          , oauthClientSecret = getClientSecret oa2ClientSecret
          , oauthOAuthorizeEndpoint = authEndpointURI
          , oauthAccessTokenEndpoint = accessTokenEndpointURI
          -- Setting callback endpoint to `Nothing` below is a lie.
          -- We do have a callback endpoint but in this context
          -- don't have access to the function that can render it.
          -- We get away with this because the callback endpoint is
          -- not needed for obtaining a refresh token, the only
          -- way we use the config here constructed.
          , oauthCallback = Nothing
          }
    man <- getGlobalManager
```

jwoudenberg replied:
Makes sense, thanks!

lehins commented on src/Network/Wai/Middleware/Auth/OAuth2.hs:
```diff
@@ -104,46 +87,58 @@ instance AuthProvider OAuth2 where
+          rRes <- refreshTokens tokens man oauth2
+          case rRes of
+            Nothing -> pure Nothing
+            Just newTokens -> 
```
Feel free to leave it as is if you like, but this is just a `fmap`, so I think it be cleaner with it, something along those lines:

```haskell
              pure (rRes <&> \ newTokens -> (req, user {
                      authLoginState = encodeToken newTokens,
                      authLoginTime = fromIntegral now
                    }))
```

jwoudenberg replied:
Yep, better. Thank you!

lehins reviewed (approved):
@jwoudenberg Sorry it took a while to respond to your PR. 

There are a couple minor improvements I suggested here, but overall looks real good. Once you take care of those suggestions feel free to rebase on master, wait for CI to pass and then release it on Hackage as 0.2.3.0.

I noticed that there was never a 0.2.2.0 release on hackage, I guess we can just skip it.

jwoudenberg commented:
Hi @lehins, thank you so much for your feedback, and sorry for taking so long to make use of it! Going through it now. Will make updates and then release as you suggested!

jwoudenberg committed:
Drop support for jose 0.7.x

jwoudenberg committed:
Add FromJSON instance to `OpenIDConnect` type

This allows it to be loaded from a configuration file

jwoudenberg committed:
Move values closer to where they're used

jwoudenberg committed:
Replace case statement with `fmap`

jwoudenberg committed:
Changelog change in supported `jose` versions

jwoudenberg committed:
Fix compilation issue from bad merge

jwoudenberg committed:
Fix CI

By default the `clientsession` library obtains an encryption key to use
for client sessions from a file. When tests run in parallel they all try
to use this file at once, leading to test flakiness.

In this commit we override the default to use a generated in memory key
instead. This should not have the same concurrency downsides.

jwoudenberg commented:
Addressed feedback, rebased on master, and CI is green! Going to merge, then release!

jwoudenberg commented:
Released as 0.2.3.0

lehins commented:
@jwoudenberg Great job! Thank you for the new release ;)
